apiserver/facades/client/charms/client.go

```diff
@@ -236,24 +230,12 @@ func normalizeCharmOrigin(origin params.CharmOrigin, fallbackArch string) (param
 func (a *API) AddCharm(args params.AddCharmWithOrigin) (params.CharmOriginResult, error) {
 	logger.Tracef("AddCharm %+v", args)
 	return a.addCharmWithAuthorization(params.AddCharmWithAuth{
-		URL:                args.URL,
-		Origin:             args.Origin,
-		CharmStoreMacaroon: nil,
-		Force:              args.Force,
+		URL:    args.URL,
+		Origin: args.Origin,
+		Force:  args.Force,
 	})
 }
 
-// AddCharmWithAuthorization adds the given charm URL (which must include
-// revision) to the environment, if it does not exist yet. Local charms are
-// not supported, only charm store and charm hub URLs. See also AddLocalCharm().
-//
-// The authorization macaroon, args.CharmStoreMacaroon, may be
-// omitted, in which case this call is equivalent to AddCharm.
-func (a *API) AddCharmWithAuthorization(args params.AddCharmWithAuth) (params.CharmOriginResult, error) {
```

Comment:
You need to rev the facade to version 6, and keep this method for version 5. juju 3.0.x clients use the method. The 3.0.x client should still be able to deploy charmhub charms to a 3.1 controller, and this method is used.

Reply:
No problem deploying a ch charm from a 3.0 client to a controller running this code.